// compress-pdf/dist/chunk-6KUCUI6Z.mjs
import {
  get_bin_path_default
} from "./chunk-ANBQPN6Q.mjs";
import {
  CompressPdfError,
  VALID_RESOLUTIONS
} from "./chunk-WPVJTEFH.mjs";

// src/compress.ts
import path from "path";
import util from "util";
import fs from "fs";
import os from "os";
import childProcess from "child_process";
import { randomUUID } from "crypto";
var execFile = util.promisify(childProcess.execFile);
function definedOptions(options) {
  if (!options) return {};
  return Object.fromEntries(
    Object.entries(options).filter((entry) => entry[1] !== void 0)
  );
}
var defaultOptions = {
  compatibilityLevel: 1.4,
  resolution: "ebook",
  imageQuality: 100,
  gsModule: get_bin_path_default(os.platform()),
  pdfPassword: "",
  removePasswordAfterCompression: false
};
function validateOptions(opts) {
  if (!VALID_RESOLUTIONS.includes(
    opts.resolution
  )) {
    throw new CompressPdfError(
      `Invalid resolution "${opts.resolution}". Must be one of: ${VALID_RESOLUTIONS.join(", ")}`
    );
  }
  if (opts.imageQuality < 1 || opts.imageQuality > 600) {
    throw new CompressPdfError(
      `imageQuality must be between 1 and 600, got ${opts.imageQuality}`
    );
  }
  if (opts.compatibilityLevel < 1 || opts.compatibilityLevel > 2) {
    throw new CompressPdfError(
      `compatibilityLevel must be between 1.0 and 2.0, got ${opts.compatibilityLevel}`
    );
  }
}
function buildGsArgs(options) {
  const args = [
    "-q",
    "-dNOPAUSE",
    "-dBATCH",
    "-dSAFER",
    "-dSimulateOverprint=true",
    "-sDEVICE=pdfwrite",
    `-dCompatibilityLevel=${options.compatibilityLevel}`,
    `-dPDFSETTINGS=/${options.resolution}`,
    "-dEmbedAllFonts=true",
    "-dSubsetFonts=true",
    "-dAutoRotatePages=/None",
    "-dColorImageDownsampleType=/Bicubic",
    `-dColorImageResolution=${options.imageQuality}`,
    "-dGrayImageDownsampleType=/Bicubic",
    `-dGrayImageResolution=${options.imageQuality}`,
    "-dMonoImageDownsampleType=/Bicubic",
    `-dMonoImageResolution=${options.imageQuality}`,
    `-sOutputFile=${options.output}`
  ];
  if (options.pdfPassword) {
    args.push(`-sPDFPassword=${options.pdfPassword}`);
  }
  if (options.pdfPassword && !options.removePasswordAfterCompression) {
    args.push(
      `-sOwnerPassword=${options.pdfPassword}`,
      `-sUserPassword=${options.pdfPassword}`
    );
  }
  args.push(options.inputFile);
  return args;
}
async function safeUnlink(filePath) {
  try {
    await fs.promises.unlink(filePath);
  } catch {
  }
}
async function compress(file, options) {
  const startTime = Date.now();
  const mergedOptions = {
    ...defaultOptions,
    ...definedOptions(options)
  };
  validateOptions(mergedOptions);
  const {
    resolution,
    imageQuality,
    compatibilityLevel,
    gsModule,
    pdfPassword,
    removePasswordAfterCompression
  } = mergedOptions;
  if (typeof file === "string" && !fs.existsSync(file)) {
    throw new CompressPdfError(`File not found: ${file}`);
  }
  const output = path.resolve(os.tmpdir(), `compress-pdf-${randomUUID()}`);
  let tempFile;
  try {
    let inputFile;
    if (typeof file === "string") {
      inputFile = file;
    } else {
      tempFile = path.resolve(os.tmpdir(), `compress-pdf-${randomUUID()}`);
      await fs.promises.writeFile(tempFile, file);
      inputFile = tempFile;
    }
    const args = buildGsArgs({
      output,
      inputFile,
      compatibilityLevel,
      resolution,
      imageQuality,
      pdfPassword,
      removePasswordAfterCompression
    });
    try {
      await execFile(gsModule, args);
    } catch (error) {
      throw new CompressPdfError(
        `Ghostscript failed to compress the PDF. ${error instanceof Error ? error.message : String(error)}`,
        error
      );
    }
    const compressedBuffer = await fs.promises.readFile(output);
    const originalSize = typeof file === "string" ? (await fs.promises.stat(file)).size : file.length;
    const duration = Date.now() - startTime;
    const compressedSize = compressedBuffer.length;
    const compressionRatio = originalSize > 0 ? compressedSize / originalSize : 0;
    Object.defineProperties(compressedBuffer, {
      buffer: { value: compressedBuffer, enumerable: false },
      originalSize: { value: originalSize, enumerable: false },
      compressedSize: { value: compressedSize, enumerable: false },
      compressionRatio: { value: compressionRatio, enumerable: false },
      duration: { value: duration, enumerable: false }
    });
    return compressedBuffer;
  } finally {
    if (tempFile) await safeUnlink(tempFile);
    await safeUnlink(output);
  }
}
var compress_default = compress;

export {
  compress_default
};
