// compress-pdf/dist/chunk-PMJIU2OQ.mjs
import {
  compress_default
} from "./chunk-6KUCUI6Z.mjs";

// src/cli.ts
import { parseArgs } from "util";
import fs from "fs";
var helpText = `
compress-pdf - Compress PDF files using Ghostscript

Usage:
  npx compress-pdf --file <input> --output <output> [options]

Required:
  -f, --file <path>          Path to the PDF file to compress
  -o, --output <path>        Path to save the compressed PDF

Options:
  -r, --resolution <preset>  Compression preset: screen | ebook | printer | prepress | default
                             (default: ebook)
  --compatibilityLevel <n>   PDF compatibility level (default: 1.4)
  --imageQuality <n>         Image resolution/quality in DPI, 1-600 (default: 100)
  --gsModule <path>          Custom Ghostscript binary path
  --pdfPassword <pass>       Password for protected PDFs
  --removePasswordAfterCompression
                             Remove password protection after compression
  -h, --help                 Show this help message

Examples:
  npx compress-pdf -f input.pdf -o output.pdf
  npx compress-pdf -f input.pdf -o output.pdf -r screen
  npx compress-pdf -f input.pdf -o output.pdf --imageQuality 72
  npx compress-pdf -f protected.pdf -o output.pdf --pdfPassword mypass
`;
function getStringValue(value) {
  return typeof value === "string" ? value : void 0;
}
async function runCli(userArgs) {
  const { values } = parseArgs({
    args: [...userArgs],
    options: {
      file: { type: "string", short: "f" },
      output: { type: "string", short: "o" },
      resolution: { type: "string", short: "r" },
      compatibilityLevel: { type: "string" },
      imageQuality: { type: "string" },
      gsModule: { type: "string" },
      pdfPassword: { type: "string" },
      removePasswordAfterCompression: { type: "boolean", default: false },
      help: { type: "boolean", short: "h", default: false }
    },
    strict: false
  });
  if (values.help || userArgs.length === 0) {
    console.log(helpText);
    return 0;
  }
  const file = getStringValue(values.file);
  const output = getStringValue(values.output);
  const resolution = getStringValue(values.resolution);
  const compatibilityLevel = getStringValue(values.compatibilityLevel);
  const imageQuality = getStringValue(values.imageQuality);
  const gsModule = getStringValue(values.gsModule);
  const pdfPassword = getStringValue(values.pdfPassword);
  if (!file || !output) {
    console.error(
      "Error: --file and --output are required.\n\nRun with --help for usage information."
    );
    return 1;
  }
  if (!fs.existsSync(file)) {
    console.error(`Error: File not found: ${file}`);
    return 1;
  }
  try {
    const result = await compress_default(file, {
      resolution: resolution ? resolution : void 0,
      compatibilityLevel: compatibilityLevel ? Number(compatibilityLevel) : void 0,
      imageQuality: imageQuality ? Number(imageQuality) : void 0,
      gsModule,
      pdfPassword,
      removePasswordAfterCompression: values.removePasswordAfterCompression
    });
    fs.writeFileSync(output, result);
    const ratio = ((1 - result.compressionRatio) * 100).toFixed(1);
    const originalKB = (result.originalSize / 1024).toFixed(1);
    const compressedKB = (result.compressedSize / 1024).toFixed(1);
    console.log(`\u2705 PDF compressed successfully!`);
    console.log(`   ${originalKB} KB \u2192 ${compressedKB} KB (${ratio}% smaller)`);
    console.log(`   Time: ${result.duration}ms`);
    console.log(`   Output: ${output}`);
    return 0;
  } catch (error) {
    console.error(
      `\u274C Compression failed: ${error instanceof Error ? error.message : String(error)}`
    );
    return 1;
  }
}

export {
  helpText,
  runCli
};
