// compress-pdf/dist/chunk-RNJGNBV4.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _chunk4N7ZB2J7js = require('./chunk-4N7ZB2J7.js');



var _chunk5IWTWCTAjs = require('./chunk-5IWTWCTA.js');

// src/compress.ts
var _path = require('path'); var _path2 = _interopRequireDefault(_path);
var _util = require('util'); var _util2 = _interopRequireDefault(_util);
var _fs = require('fs'); var _fs2 = _interopRequireDefault(_fs);
var _os = require('os'); var _os2 = _interopRequireDefault(_os);
var _child_process = require('child_process'); var _child_process2 = _interopRequireDefault(_child_process);
var _crypto = require('crypto');
var execFile = _util2.default.promisify(_child_process2.default.execFile);
function definedOptions(options) {
  if (!options) return {};
  return Object.fromEntries(
    Object.entries(options).filter((entry) => entry[1] !== void 0)
  );
}
var defaultOptions = {
  compatibilityLevel: 1.4,
  resolution: "ebook",
  imageQuality: 100,
  gsModule: _chunk4N7ZB2J7js.get_bin_path_default.call(void 0, _os2.default.platform()),
  pdfPassword: "",
  removePasswordAfterCompression: false
};
function validateOptions(opts) {
  if (!_chunk5IWTWCTAjs.VALID_RESOLUTIONS.includes(
    opts.resolution
  )) {
    throw new (0, _chunk5IWTWCTAjs.CompressPdfError)(
      `Invalid resolution "${opts.resolution}". Must be one of: ${_chunk5IWTWCTAjs.VALID_RESOLUTIONS.join(", ")}`
    );
  }
  if (opts.imageQuality < 1 || opts.imageQuality > 600) {
    throw new (0, _chunk5IWTWCTAjs.CompressPdfError)(
      `imageQuality must be between 1 and 600, got ${opts.imageQuality}`
    );
  }
  if (opts.compatibilityLevel < 1 || opts.compatibilityLevel > 2) {
    throw new (0, _chunk5IWTWCTAjs.CompressPdfError)(
      `compatibilityLevel must be between 1.0 and 2.0, got ${opts.compatibilityLevel}`
    );
  }
}
function buildGsArgs(options) {
  const args = [
    "-q",
    "-dNOPAUSE",
    "-dBATCH",
    "-dSAFER",
    "-dSimulateOverprint=true",
    "-sDEVICE=pdfwrite",
    `-dCompatibilityLevel=${options.compatibilityLevel}`,
    `-dPDFSETTINGS=/${options.resolution}`,
    "-dEmbedAllFonts=true",
    "-dSubsetFonts=true",
    "-dAutoRotatePages=/None",
    "-dColorImageDownsampleType=/Bicubic",
    `-dColorImageResolution=${options.imageQuality}`,
    "-dGrayImageDownsampleType=/Bicubic",
    `-dGrayImageResolution=${options.imageQuality}`,
    "-dMonoImageDownsampleType=/Bicubic",
    `-dMonoImageResolution=${options.imageQuality}`,
    `-sOutputFile=${options.output}`
  ];
  if (options.pdfPassword) {
    args.push(`-sPDFPassword=${options.pdfPassword}`);
  }
  if (options.pdfPassword && !options.removePasswordAfterCompression) {
    args.push(
      `-sOwnerPassword=${options.pdfPassword}`,
      `-sUserPassword=${options.pdfPassword}`
    );
  }
  args.push(options.inputFile);
  return args;
}
async function safeUnlink(filePath) {
  try {
    await _fs2.default.promises.unlink(filePath);
  } catch (e) {
  }
}
async function compress(file, options) {
  const startTime = Date.now();
  const mergedOptions = {
    ...defaultOptions,
    ...definedOptions(options)
  };
  validateOptions(mergedOptions);
  const {
    resolution,
    imageQuality,
    compatibilityLevel,
    gsModule,
    pdfPassword,
    removePasswordAfterCompression
  } = mergedOptions;
  if (typeof file === "string" && !_fs2.default.existsSync(file)) {
    throw new (0, _chunk5IWTWCTAjs.CompressPdfError)(`File not found: ${file}`);
  }
  const output = _path2.default.resolve(_os2.default.tmpdir(), `compress-pdf-${_crypto.randomUUID.call(void 0, )}`);
  let tempFile;
  try {
    let inputFile;
    if (typeof file === "string") {
      inputFile = file;
    } else {
      tempFile = _path2.default.resolve(_os2.default.tmpdir(), `compress-pdf-${_crypto.randomUUID.call(void 0, )}`);
      await _fs2.default.promises.writeFile(tempFile, file);
      inputFile = tempFile;
    }
    const args = buildGsArgs({
      output,
      inputFile,
      compatibilityLevel,
      resolution,
      imageQuality,
      pdfPassword,
      removePasswordAfterCompression
    });
    try {
      await execFile(gsModule, args);
    } catch (error) {
      throw new (0, _chunk5IWTWCTAjs.CompressPdfError)(
        `Ghostscript failed to compress the PDF. ${error instanceof Error ? error.message : String(error)}`,
        error
      );
    }
    const compressedBuffer = await _fs2.default.promises.readFile(output);
    const originalSize = typeof file === "string" ? (await _fs2.default.promises.stat(file)).size : file.length;
    const duration = Date.now() - startTime;
    const compressedSize = compressedBuffer.length;
    const compressionRatio = originalSize > 0 ? compressedSize / originalSize : 0;
    Object.defineProperties(compressedBuffer, {
      buffer: { value: compressedBuffer, enumerable: false },
      originalSize: { value: originalSize, enumerable: false },
      compressedSize: { value: compressedSize, enumerable: false },
      compressionRatio: { value: compressionRatio, enumerable: false },
      duration: { value: duration, enumerable: false }
    });
    return compressedBuffer;
  } finally {
    if (tempFile) await safeUnlink(tempFile);
    await safeUnlink(output);
  }
}
var compress_default = compress;



exports.compress_default = compress_default;
